File: examples/python/v2_flight_services/raster.py
"""Lawnmower raster geometry for the WUAS search service.

Pure geometry, no NDN dependencies: the drone agent flies these legs and
the dashboard renders exactly the same ones before the operator commits,
so what you preview is what flies.

Conventions: axis-aligned rectangles in a local flat-earth frame around
the area center (fine at search scales of tens to hundreds of meters);
legs run along the rectangle's LONGER axis (fewest turns), serpentine
ordering; capture points are spaced along each leg and inherit the leg's
heading so the GCS can geo-project detections.

Smoke test:
    python3 raster.py            # sample area -> legs/captures JSON
"""

from __future__ import annotations

import math
from dataclasses import dataclass
from typing import Any

from contracts import SearchArea

EARTH_M_PER_DEG_LAT = 111_111.0


def m_per_deg_lon(lat_deg: float) -> float:
    return EARTH_M_PER_DEG_LAT * max(math.cos(math.radians(lat_deg)), 1e-6)


@dataclass(frozen=True)
class CapturePoint:
    lat_deg: float
    lon_deg: float
    heading_deg: float
    leg: int
    index: int

    def as_dict(self) -> dict[str, Any]:
        return {
            "lat": self.lat_deg,
            "lon": self.lon_deg,
            "heading_deg": self.heading_deg,
            "leg": self.leg,
            "index": self.index,
        }


@dataclass(frozen=True)
class RasterPlan:
    """Resolved raster: rectangle corners, serpentine legs, capture points."""

    center_lat: float
    center_lon: float
    width_m: float   # east-west extent
    height_m: float  # north-south extent
    legs: list[list[tuple[float, float]]]          # [(lat, lon) start, end]
    captures: list[CapturePoint]

    @property
    def corners(self) -> list[tuple[float, float]]:
        """NW, NE, SE, SW (lat, lon) — render-ready polygon order."""
        dlat = (self.height_m / 2.0) / EARTH_M_PER_DEG_LAT
        dlon = (self.width_m / 2.0) / m_per_deg_lon(self.center_lat)
        n, s = self.center_lat + dlat, self.center_lat - dlat
        w, e = self.center_lon - dlon, self.center_lon + dlon
        return [(n, w), (n, e), (s, e), (s, w)]

    def as_dict(self) -> dict[str, Any]:
        return {
            "center": {"lat": self.center_lat, "lon": self.center_lon},
            "width_m": self.width_m,
            "height_m": self.height_m,
            "corners": [{"lat": a, "lon": b} for a, b in self.corners],
            "legs": [
                [{"lat": a, "lon": b} for a, b in leg] for leg in self.legs
            ],
            "captures": [c.as_dict() for c in self.captures],
        }


def resolve_area(area: SearchArea) -> tuple[float, float, float, float]:
    """SearchArea (either mode) -> (center_lat, center_lon, width_m, height_m)."""

    if area.mode == "corners" and len(area.corner_a) == 2 and len(area.corner_b) == 2:
        lat_a, lon_a = area.corner_a
        lat_b, lon_b = area.corner_b
        center_lat = (lat_a + lat_b) / 2.0
        center_lon = (lon_a + lon_b) / 2.0
        height_m = abs(lat_a - lat_b) * EARTH_M_PER_DEG_LAT
        width_m = abs(lon_a - lon_b) * m_per_deg_lon(center_lat)
        return center_lat, center_lon, width_m, height_m
    return area.center_lat, area.center_lon, area.width_m, area.height_m


def build_raster(
    area: SearchArea,
    *,
    leg_spacing_m: float = 5.0,
    capture_every_m: float = 4.0,
) -> RasterPlan:
    """Serpentine lawnmower covering the rectangle.

    Legs run along the longer axis; spacing is shrunk slightly when needed
    so legs cover the full perpendicular extent symmetrically (first and
    last legs sit half a spacing inside the boundary, matching the camera
    footprint center, not its edge).
    """

    center_lat, center_lon, width_m, height_m = resolve_area(area)
    leg_spacing_m = max(float(leg_spacing_m), 0.5)
    capture_every_m = max(float(capture_every_m), 0.5)

    along_ew = width_m >= height_m  # legs run east-west when wider than tall
    along_m = width_m if along_ew else height_m
    across_m = height_m if along_ew else width_m

    leg_count = max(1, int(math.ceil(across_m / leg_spacing_m)))
    # symmetric offsets across the perpendicular axis, centered on 0
    if leg_count == 1:
        across_offsets = [0.0]
    else:
        step = across_m / leg_count
        start = -(across_m / 2.0) + step / 2.0
        across_offsets = [start + i * step for i in range(leg_count)]

    half_along = along_m / 2.0
    lat_per_m = 1.0 / EARTH_M_PER_DEG_LAT
    lon_per_m = 1.0 / m_per_deg_lon(center_lat)

    def to_latlon(along: float, across: float) -> tuple[float, float]:
        if along_ew:
            east, north = along, across
        else:
            east, north = across, along
        return (
            center_lat + north * lat_per_m,
            center_lon + east * lon_per_m,
        )

    legs: list[list[tuple[float, float]]] = []
    captures: list[CapturePoint] = []
    for leg_index, across in enumerate(across_offsets):
        forward = leg_index % 2 == 0  # serpentine
        a, b = (-half_along, half_along) if forward else (half_along, -half_along)
        start = to_latlon(a, across)
        end = to_latlon(b, across)
        legs.append([start, end])

        if along_ew:
            heading = 90.0 if forward else 270.0
        else:
            heading = 0.0 if forward else 180.0

        n_caps = max(1, int(math.floor(along_m / capture_every_m)) + 1)
        for cap_index in range(n_caps):
            frac = (cap_index / (n_caps - 1)) if n_caps > 1 else 0.5
            along_pos = a + (b - a) * frac
            lat, lon = to_latlon(along_pos, across)
            captures.append(
                CapturePoint(
                    lat_deg=lat,
                    lon_deg=lon,
                    heading_deg=heading,
                    leg=leg_index,
                    index=len(captures),
                )
            )

    return RasterPlan(
        center_lat=center_lat,
        center_lon=center_lon,
        width_m=width_m,
        height_m=height_m,
        legs=legs,
        captures=captures,
    )


def estimate_duration_s(
    plan: RasterPlan, *, speed_m_s: float, turn_penalty_s: float = 3.0
) -> float:
    """Coarse flight-time estimate for UI display."""

    along = max(plan.width_m, plan.height_m)
    legs = len(plan.legs)
    across_travel = min(plan.width_m, plan.height_m)
    distance = legs * along + across_travel
    return distance / max(speed_m_s, 0.1) + max(legs - 1, 0) * turn_penalty_s


if __name__ == "__main__":
    import json

    sample = SearchArea(
        mode="center",
        center_lat=35.1208,
        center_lon=-89.9347,
        width_m=40.0,
        height_m=24.0,
    )
    plan = build_raster(sample, leg_spacing_m=5.0, capture_every_m=4.0)
    print(json.dumps(
        {
            **plan.as_dict(),
            "legs_count": len(plan.legs),
            "captures_count": len(plan.captures),
            "estimate_s": round(estimate_duration_s(plan, speed_m_s=2.0), 1),
        },
        indent=2,
    ))

File: crates/muas-dashboard/src/raster.rs
//! Raster preview built on uas-flight's pattern geometry — the v3 form of
//! v2's "what you preview is what flies" rule (`raster.py` was shared by
//! agent and dashboard; v3 shares [`uas_flight::patterns::RasterPath`]).
//!
//! The preview response keeps the v2 `raster_preview` JSON shape (center,
//! corners NW/NE/SE/SW, legs, capture points, duration estimate) so the
//! ported frontend renders it unchanged.
//!
//! Documented deviation from v2 `raster.py`: lane placement follows
//! uas-flight's `lane_offsets` (lanes span the rectangle edge-to-edge,
//! `ceil(w/s)+1` lanes, minimum 2) instead of v2's half-spacing-inset lanes,
//! because that is the geometry the v3 agent flies. Legs still run along
//! the longer axis with serpentine ordering, and captures are spaced evenly
//! along each leg including both endpoints, exactly like v2.

use serde_json::{json, Value};
use uas_flight::geo::{m_per_deg_lon, Position, EARTH_M_PER_DEG_LAT};
use uas_flight::motion::MotionTarget;
use uas_flight::patterns::{RasterBounds, RasterPath};

/// A resolved preview plan (v2 `RasterPlan` facts).
#[derive(Debug, Clone, PartialEq)]
pub struct PreviewPlan {
    pub center_lat: f64,
    pub center_lon: f64,
    /// East–west extent, metres.
    pub width_m: f64,
    /// North–south extent, metres.
    pub height_m: f64,
    /// Serpentine legs as `[(lat, lon) start, (lat, lon) end]`.
    pub legs: Vec<[(f64, f64); 2]>,
    /// Capture points `(lat, lon, heading_deg, leg_index)`.
    pub captures: Vec<(f64, f64, f64, usize)>,
}

/// v2 `resolve_area`: a `SearchArea` dict (either mode) →
/// `(center_lat, center_lon, width_m, height_m)`.
pub fn resolve_area(area: &Value) -> (f64, f64, f64, f64) {
    let f = |key: &str| area.get(key).and_then(Value::as_f64).unwrap_or(0.0);
    let mode = area.get("mode").and_then(Value::as_str).unwrap_or("center");
    if mode == "corners" {
        let corner = |key: &str| -> Option<(f64, f64)> {
            let c = area.get(key)?.as_array()?;
            Some((c.first()?.as_f64()?, c.get(1)?.as_f64()?))
        };
        if let (Some((lat_a, lon_a)), Some((lat_b, lon_b))) =
            (corner("corner_a"), corner("corner_b"))
        {
            let center_lat = (lat_a + lat_b) / 2.0;
            let center_lon = (lon_a + lon_b) / 2.0;
            let height_m = (lat_a - lat_b).abs() * EARTH_M_PER_DEG_LAT;
            let width_m = (lon_a - lon_b).abs() * m_per_deg_lon(center_lat);
            return (center_lat, center_lon, width_m, height_m);
        }
    }
    (f("center_lat"), f("center_lon"), f("width_m"), f("height_m"))
}

/// Rectangle corners NW, NE, SE, SW (render-ready polygon order, v2
/// `RasterPlan.corners`).
pub fn area_corners(center_lat: f64, center_lon: f64, width_m: f64, height_m: f64) -> [(f64, f64); 4] {
    let dlat = (height_m / 2.0) / EARTH_M_PER_DEG_LAT;
    let dlon = (width_m / 2.0) / m_per_deg_lon(center_lat);
    let (n, s) = (center_lat + dlat, center_lat - dlat);
    let (w, e) = (center_lon - dlon, center_lon + dlon);
    [(n, w), (n, e), (s, e), (s, w)]
}

/// Corners for a v3 `RasterRequest` from a v2 area dict.
pub fn corners_for_area(area: &Value) -> Vec<(f64, f64)> {
    let (clat, clon, w, h) = resolve_area(area);
    area_corners(clat, clon, w, h).to_vec()
}

/// The uas-flight raster path for an area — the SAME structure the agent
/// flies. Legs run along the longer axis (fewest turns): east–west when the
/// area is wider than tall (bearing 90°), north–south otherwise (0°).
pub fn flight_path(
    center_lat: f64,
    center_lon: f64,
    width_m: f64,
    height_m: f64,
    leg_spacing_m: f64,
) -> RasterPath {
    let along_ew = width_m >= height_m;
    let (length_m, across_m, bearing_deg) = if along_ew {
        (width_m.max(0.5), height_m.max(0.5), 90.0)
    } else {
        (height_m.max(0.5), width_m.max(0.5), 0.0)
    };
    let mut path = RasterPath::new(RasterBounds {
        center: Position::new(center_lat, center_lon, 0.0),
        length_m,
        width_m: across_m,
        bearing_deg,
    });
    path.lane_spacing_m = leg_spacing_m.max(0.5);
    path
}

/// Build the preview plan by pairing the flight path's lane-endpoint
/// targets into legs and spacing captures along each leg (v2 capture rule:
/// `max(1, floor(len/step)+1)` points, endpoints included).
pub fn build_preview(
    area: &Value,
    leg_spacing_m: f64,
    capture_every_m: f64,
) -> Result<PreviewPlan, String> {
    let (center_lat, center_lon, width_m, height_m) = resolve_area(area);
    let path = flight_path(center_lat, center_lon, width_m, height_m, leg_spacing_m);
    let targets: Vec<MotionTarget> = path.targets().map_err(|e| e.to_string())?;
    let capture_every_m = capture_every_m.max(0.5);

    let mut legs = Vec::new();
    let mut captures = Vec::new();
    for (leg_index, pair) in targets.chunks(2).enumerate() {
        let [a, b] = pair else { continue };
        let start = (a.position.lat, a.position.lon);
        let end = (b.position.lat, b.position.lon);
        legs.push([start, end]);

        // Leg heading from the travel direction (flat-earth bearing).
        let dn = (end.0 - start.0) * EARTH_M_PER_DEG_LAT;
        let de = (end.1 - start.1) * m_per_deg_lon(center_lat);
        let heading = de.atan2(dn).to_degrees().rem_euclid(360.0);
        let leg_len = dn.hypot(de);

        let n_caps = usize::max(1, (leg_len / capture_every_m).floor() as usize + 1);
        for i in 0..n_caps {
            let frac = if n_caps > 1 {
                i as f64 / (n_caps - 1) as f64
            } else {
                0.5
            };
            captures.push((
                start.0 + (end.0 - start.0) * frac,
                start.1 + (end.1 - start.1) * frac,
                heading,
                leg_index,
            ));
        }
    }
    Ok(PreviewPlan {
        center_lat,
        center_lon,
        width_m,
        height_m,
        legs,
        captures,
    })
}

/// v2 `estimate_duration_s`: coarse flight-time estimate for UI display.
pub fn estimate_duration_s(plan: &PreviewPlan, speed_m_s: f64) -> f64 {
    const TURN_PENALTY_S: f64 = 3.0;
    let along = plan.width_m.max(plan.height_m);
    let legs = plan.legs.len() as f64;
    let across_travel = plan.width_m.min(plan.height_m);
    let distance = legs * along + across_travel;
    distance / speed_m_s.max(0.1) + (legs - 1.0).max(0.0) * TURN_PENALTY_S
}

/// The full `raster_preview` message (v2 wire shape).
pub fn preview_message(area: &Value, leg_spacing_m: f64, capture_every_m: f64, speed_m_s: f64) -> Value {
    match build_preview(area, leg_spacing_m, capture_every_m) {
        Ok(plan) => {
            let estimate = estimate_duration_s(&plan, speed_m_s);
            json!({
                "type": "raster_preview",
                "plan": plan_to_json(&plan),
                "estimate_s": (estimate * 10.0).round() / 10.0,
            })
        }
        Err(error) => json!({ "type": "raster_preview", "error": error }),
    }
}

/// v2 `RasterPlan.as_dict`.
pub fn plan_to_json(plan: &PreviewPlan) -> Value {
    json!({
        "center": { "lat": plan.center_lat, "lon": plan.center_lon },
        "width_m": plan.width_m,
        "height_m": plan.height_m,
        "corners": area_corners(plan.center_lat, plan.center_lon, plan.width_m, plan.height_m)
            .iter()
            .map(|(lat, lon)| json!({ "lat": lat, "lon": lon }))
            .collect::<Vec<_>>(),
        "legs": plan
            .legs
            .iter()
            .map(|leg| {
                json!([
                    { "lat": leg[0].0, "lon": leg[0].1 },
                    { "lat": leg[1].0, "lon": leg[1].1 },
                ])
            })
            .collect::<Vec<_>>(),
        "captures": plan
            .captures
            .iter()
            .enumerate()
            .map(|(index, (lat, lon, heading, leg))| {
                json!({
                    "lat": lat,
                    "lon": lon,
                    "heading_deg": heading,
                    "leg": leg,
                    "index": index,
                })
            })
            .collect::<Vec<_>>(),
    })
}
